File: Supplementary_webpages/static/gallery_res/Polygon Mesh Processing page 42/pmp_42.py
"""
`n(v)` = (∑_(i for i ∈ `N₁(v)`) α_i n(T_i))/‖∑_(i for i ∈ `N₁(v)`) α_i n(T_i)‖

where
 
T_i ∈ ℝ^(3×3)
α_i ∈ ℝ
`N₁(v)` ∈ {ℤ}
n: ℝ^(3×3) → ℝ^3
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


class pmp_42ResultType:
    def __init__( self, n_left_parenthesis_v_right_parenthesis):
        self.n_left_parenthesis_v_right_parenthesis = n_left_parenthesis_v_right_parenthesis


def pmp_42(T, α, N1_left_parenthesis_v_right_parenthesis, n):
    """
    :param :n : ℝ^(3×3) → ℝ^3
    """
    T = np.asarray(T, dtype=np.float64)
    α = np.asarray(α, dtype=np.float64)

    dim_0 = α.shape[0]
    assert T.shape == (dim_0, 3, 3)
    assert α.shape == (dim_0,)
    assert isinstance(N1_left_parenthesis_v_right_parenthesis, list) and len(N1_left_parenthesis_v_right_parenthesis) > 0

    sum_0 = np.zeros((3, ))
    for i in range(1, len(α)+1):
        if((i) in N1_left_parenthesis_v_right_parenthesis):
            sum_0 += α[i-1] * n(T[i-1])
    sum_1 = np.zeros((3, ))
    for i in range(1, len(α)+1):
        if((i) in N1_left_parenthesis_v_right_parenthesis):
            sum_1 += α[i-1] * n(T[i-1])
    n_left_parenthesis_v_right_parenthesis = (sum_0) / np.linalg.norm(sum_1, 2)
    return pmp_42ResultType(n_left_parenthesis_v_right_parenthesis)


def generateRandomData():
    dim_0 = np.random.randint(10)
    T = np.random.randn(dim_0, 3, 3)
    α = np.random.randn(dim_0)
    N1_left_parenthesis_v_right_parenthesis = []
    dim_1 = np.random.randint(1, 10)
    for i in range(dim_1):
        N1_left_parenthesis_v_right_parenthesis.append((np.random.randint(10)))
    def n(p0):
        return np.random.randn(3)
    return T, α, N1_left_parenthesis_v_right_parenthesis, n


if __name__ == '__main__':
    T, α, N1_left_parenthesis_v_right_parenthesis, n = generateRandomData()
    print("T:", T)
    print("α:", α)
    print("N1_left_parenthesis_v_right_parenthesis:", N1_left_parenthesis_v_right_parenthesis)
    print("n:", n)
    func_value = pmp_42(T, α, N1_left_parenthesis_v_right_parenthesis, n)
    print("return value: ", func_value.n_left_parenthesis_v_right_parenthesis)

File: Supplementary_webpages/static/gallery_res/Polygon Mesh Processing page 42/pmp_42.cpp
/*
`n(v)` = (∑_(i for i ∈ `N₁(v)`) α_i n(T_i))/‖∑_(i for i ∈ `N₁(v)`) α_i n(T_i)‖

where
 
T_i ∈ ℝ^(3×3)
α_i ∈ ℝ
`N₁(v)` ∈ {ℤ}
n: ℝ^(3×3) → ℝ^3
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

struct pmp_42ResultType {
    Eigen::Matrix<double, 3, 1> n_left_parenthesis_v_right_parenthesis;
    pmp_42ResultType(const Eigen::Matrix<double, 3, 1> & n_left_parenthesis_v_right_parenthesis)
    : n_left_parenthesis_v_right_parenthesis(n_left_parenthesis_v_right_parenthesis)
    {}
};

/**
 * pmp_42
 *
 * @param n  ℝ^(3×3) → ℝ^3
 * @return n_left_parenthesis_v_right_parenthesis
 */
pmp_42ResultType pmp_42(
    const std::vector<Eigen::Matrix<double, 3, 3>> & T,
    const std::vector<double> & α,
    const std::set<std::tuple< int > > & N₁_left_parenthesis_v_right_parenthesis,
    const std::function<Eigen::Matrix<double, 3, 1>(Eigen::Matrix<double, 3, 3>)> & n)
{
    const long dim_0 = α.size();
    assert( T.size() == dim_0 );

    Eigen::MatrixXd sum_0 = Eigen::MatrixXd::Zero(3, 1);
    for(int i=1; i<=α.size(); i++){
        if(N₁_left_parenthesis_v_right_parenthesis.find(std::tuple< int >(i)) != N₁_left_parenthesis_v_right_parenthesis.end()){
            sum_0 += α.at(i-1) * n(T.at(i-1));
        }
    }
    Eigen::MatrixXd sum_1 = Eigen::MatrixXd::Zero(3, 1);
    for(int i=1; i<=α.size(); i++){
        if(N₁_left_parenthesis_v_right_parenthesis.find(std::tuple< int >(i)) != N₁_left_parenthesis_v_right_parenthesis.end()){
            sum_1 += α.at(i-1) * n(T.at(i-1));
        }
    }
    Eigen::Matrix<double, 3, 1> n_left_parenthesis_v_right_parenthesis = (sum_0) / double((sum_1).lpNorm<2>());

    return pmp_42ResultType(n_left_parenthesis_v_right_parenthesis);
}


void generateRandomData(std::vector<Eigen::Matrix<double, 3, 3>> & T,
    std::vector<double> & α,
    std::set<std::tuple< int > > & N₁_left_parenthesis_v_right_parenthesis,
    std::function<Eigen::Matrix<double, 3, 1>(Eigen::Matrix<double, 3, 3>)> & n)
{
    const int dim_0 = rand()%10;
    T.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        T[i] = Eigen::MatrixXd::Random(3, 3);
    }
    α.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        α[i] = rand() % 10;
    }
    const int dim_2 = rand()%10;
    for(int i=0; i<dim_2; i++){
        N₁_left_parenthesis_v_right_parenthesis.insert(std::make_tuple(rand()%10));
    }
    n = [](Eigen::Matrix<double, 3, 3>)->Eigen::Matrix<double, 3, 1>{
        return Eigen::VectorXd::Random(3);
    };
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    std::vector<Eigen::Matrix<double, 3, 3>> T;
    std::vector<double> α;
    std::set<std::tuple< int > > N₁_left_parenthesis_v_right_parenthesis;
    std::function<Eigen::Matrix<double, 3, 1>(Eigen::Matrix<double, 3, 3>)> n;
    generateRandomData(T, α, N₁_left_parenthesis_v_right_parenthesis, n);
    pmp_42ResultType func_value = pmp_42(T, α, N₁_left_parenthesis_v_right_parenthesis, n);
    std::cout<<"return value:\n"<<func_value.n_left_parenthesis_v_right_parenthesis<<std::endl;
    return 0;
}